pennylane/plugins/expt_tensornet.py

```diff
@@ -0,0 +1,197 @@
+        self._edges.append(edge)
+        return edge
+
+    def pre_apply(self):
+        self.reset()
+
+    def apply(self, operation, wires, par):
+        A = self._get_operator_matrix(operation, par)
+        num_mult_idxs = len(wires)
+        A = np.reshape(A, [2] * num_mult_idxs * 2)
+        op_node = self._add_node(A, wires=wires, name=operation)
+        for idx, w in enumerate(wires):
+            self._add_edge(op_node, num_mult_idxs + idx, self._state, w)
+            self._free_edges[w] = op_node[idx]
+        # TODO: can be smarter here about collecting contractions?
+        self._state = tn.contract_between(op_node, self._state, output_edge_order=self._free_edges)
+        print(self._state._tensor.reshape(-1))
+        print("\n")
+
+    def expval(self, observable, wires, par):
+        if not isinstance(observable, list):
+            observable = [observable]
+            wires = [wires]
+            par = [par]
+        matrices = []
+        for o, p, w in zip(observable, par, wires):
+            A = self._get_operator_matrix(o, p)
+            num_mult_idxs = len(w)
+            matrices.append(np.reshape(A, [2] * num_mult_idxs * 2))
+        nodes = [self._add_node(A, w, name=o) for A, w, o in zip(matrices, wires, observable)]
+        return self.ev(nodes, wires)
+
+    def _get_operator_matrix(self, operation, par):
+        """Get the operator matrix for a given operation or observable.
+
+        Args:
+          operation    (str): name of the operation/observable
+          par (tuple[float]): parameter values
+        Returns:
+          array: matrix representation.
+        """
+        A = {**self._operation_map, **self._observable_map}[operation]
+        if not callable(A):
+            return A
+        return A(*par)
+
+    def ev(self, obs_nodes, wires):
+        r"""Expectation value of observables on specified wires.
+
+         Args:
+            obs_nodes (tn.Node): the observables as tensornetwork Nodes
+            wires (Sequence[int]): measured subsystems for each observable
+         Returns:
+            float: expectation value :math:`\expect{A} = \bra{\psi}A\ket{\psi}`
+        """
+        # first need to connect together nodes representing measurement
+        all_wires = tuple(w for w in range(self.num_wires))
+        ket = self._add_node(self._state, wires=all_wires, name="Ket")
+        bra = self._add_node(tn.conj(ket), wires=all_wires, name="Bra")
+        meas_wires = []
+        for obs_node, obs_wires in zip(obs_nodes, wires):
+            meas_wires.extend(obs_wires)
+            for idx, w in enumerate(obs_wires):
+                left_mult_index = idx
+                right_mult_index = len(obs_wires) + idx
+                self._add_edge(obs_node, right_mult_index, ket, w)  # A|psi>
+                self._add_edge(bra, w, obs_node, left_mult_index)  # <psi|A
+        for w in set(all_wires) - set(meas_wires):
+            self._add_edge(bra, w, ket, w)  # |psi[w]|**2
```

Comment:
This is really nice, but might be improved with line comments! Took me a while to go through and verify

Reply:
You mean just line 175?

Reply:
Should have used a multi-line comment, but most of the block (particularly right_mult_index and left_mult_index)